common/component/postgresql/databases/databases.go

```diff
@@ -0,0 +1,91 @@
+package databases
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pginterfaces "github.com/dapr/components-contrib/common/component/postgresql/interfaces"
+	"github.com/jackc/pgx/v5"
+)
+
+const (
+	createDatabaseTmpl = "CREATE DATABASE %s"
+	deleteDatabaseTmpl = "DROP DATABASE IF EXISTS %s WITH (FORCE)"
+	databaseExistsTmpl = "SELECT EXISTS (SELECT 1 FROM pg_database WHERE datname = $1)"
+	userExistsTmpl     = "SELECT CAST((SELECT 1 FROM pg_roles WHERE rolname = $1) AS BOOLEAN)"
+	createUserTmpl     = "CREATE USER %v WITH PASSWORD '%v'"
+	checkRoleTmpl      = "SELECT 1 FROM pg_roles WHERE rolname = $1 AND 'rds_iam' = rolname"
+	grantRoleTmpl      = "GRANT %v TO %v"
+
+	awsRole = "rds_iam"
+)
+
+func CreateDatabaseIfNeeded(ctx context.Context, timeout time.Duration, connectionString string, db pginterfaces.PGXPoolConn) error {
+	config, err := pgx.ParseConfig(connectionString)
+	if err != nil {
+		return fmt.Errorf("failed to parse PostgreSQL config: %w", err)
+	}
+
+	// check if database exists in connection string
+	var dbExists bool
+	dbExistsCtx, dbExistsCancel := context.WithTimeout(ctx, timeout)
+	err = db.QueryRow(dbExistsCtx, databaseExistsTmpl, config.Database).Scan(&dbExists)
+	dbExistsCancel()
+	if err != nil && err != pgx.ErrNoRows {
+		return fmt.Errorf("failed to check if the PostgreSQL database %s exists: %w", config.Database, err)
+	}
+
+	// Create database if needed using master password in connection string
+	if !dbExists {
+		createDBCtx, createDBCancel := context.WithTimeout(ctx, timeout)
+		_, err := db.Exec(createDBCtx, fmt.Sprintf(createDatabaseTmpl, config.Database))
+		createDBCancel()
+		if err != nil {
+			return fmt.Errorf("failed to create PostgreSQL user: %w", err)
+		}
+	}
+	return nil
+}
+
+func CreateUserAndRoleIfNeeded(ctx context.Context, timeout time.Duration, connectionString string, db pginterfaces.PGXPoolConn) error {
+	config, err := pgx.ParseConfig(connectionString)
+	if err != nil {
+		return fmt.Errorf("failed to parse PostgreSQL config: %w", err)
+	}
+
+	var userExists bool
+	userExistsCtx, userExistsCancel := context.WithTimeout(ctx, timeout)
+	err = db.QueryRow(userExistsCtx, userExistsTmpl, config.User).Scan(&userExists)
+	userExistsCancel()
+	if err != nil && err != pgx.ErrNoRows {
+		// If user does not exist, create the user with pwd = user name for now.
+		// This will be rotated to use proper iam credentials before connecting again.
```

Comment:
I really do not like this.

Although the code below changes the password to IAM credentials, if Dapr crashes half-way (between the time the user is created and its password is changed), you are stuck with a user that has password = user name. That is a HUGE security risk.

Reply:
addressed. thoughts on changes?

Reply:
I see the change but I am still not sure it's a good idea. Now you're creating a user without a password, and if Daprd crashes half-way (and maybe never restarts!), you are left with a user without a password. This is a MASSIVE security risk.

I would recommend removing all code that creates the database from Dapr. We should add it to the docs as a pre-requisite instead.